Remove char[] allocations from RegistryKey

stephentoub opened the pull request:
Replace char[] allocations with stackalloc'd memory, or ArrayPool as a fallback if the size is too large.

I also noticed that the handler for REG_EXPAND_SZ was identical to the handler for REG_SZ save for one block, so I just collapsed the two and added an additional condition to that one block.

jkotas commented on src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs:
```diff
@@ -702,14 +687,25 @@ private unsafe string[] GetValueNamesCore(int values)
 
                         // make sure the string is null terminated before processing the data
-                        if (blob.Length > 0 && blob[blob.Length - 1] != (char)0)
+                        if (blob.Length > 0 && blob[blobLength - 1] != '\0')
```
If the registry value chanced between the two `RegQueryValueEx` calls, the blob can contain uninitialized data now (it is not zero-filled anymore) and this code can have unpredictable behavior.

jkotas commented on src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs:
```diff
@@ -637,57 +639,40 @@ private unsafe string[] GetValueNamesCore(int values)
-                        }
-                        else
+                        char[]? pooledArray = null;
+                        int blobLength = datasize / 2;
```
Now that this is using stackallocation and pooled buffer, it would be nice to avoid the two RegQueryValueEx calls for common cases. We can stackallocate the buffer and pass it into the first RegQueryValueEx call. If the buffer was big enough, we are done. If the buffer was not big enough, we retry with properly sized buffer.

stephentoub replied:
In making this change, we appear to not properly handle the cases where the key is changed concurrently with the operation, e.g. there's no retry for REG_BINARY.  Is there a reason that's not needed?

stephentoub replied:
Ok, I'm just going to rewrite this method.

stephentoub replied:
@jkotas, we have a handful of tests that are trying to validate roundtripping of corrupted data, e.g. storing new byte[] { 6, 5, 6 } into a REG_SZ key and expecting to get back a string containing { 0x0506, 6 }.  But RegQueryValueEx is showing some interesting behaviors here.  If you pass in a buffer length of 3 or 4 to retrieve that data, you get it back, but any buffer length larger than 4, and the 3rd byte is zero'd such that we instead get back just { 0x0506 }.  As such I don't see a good way to retain the existing behavior in these (presumably all corrupt) cases while also supporting making a single call to RegQueryValueEx.  I'm tempted to say it's an acceptable change: corrupt in, corrupt out.  Opinions?

bartonjs commented on src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs:
```diff
@@ -637,57 +639,41 @@ private unsafe string[] GetValueNamesCore(int values)
-                            // in the very unlikely case the data is missing null termination,
-                            // pass in the whole char[] to prevent truncating a character
-                            data = new string(blob);
+                            Interop.Advapi32.RegQueryValueEx(_hkey, name, null, ref type, blobPtr, ref datasize);
```
I'm surprised we don't care about the return value...

bartonjs commented on src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs:
```diff
@@ -637,57 +639,41 @@ private unsafe string[] GetValueNamesCore(int values)
-                case Interop.Advapi32.RegistryValues.REG_EXPAND_SZ:
-                    {
-                        if (datasize % 2 == 1)
+                        if (!blob.IsEmpty && blob[blobLength - 1] == '\0')
```
This routine seems to take it for granted that datasize in the first call was accurate.  Maybe registry does guarantee that, but not all Win32 does (at least some of Win32 Crypto reports "here's a number that's guaranteed big enough, but the answer might be smaller when you make me do actual work").  Seems like this would be more robust if it recomputed based on `datasize` after the call.

But maybe it's fine.  After all, we've had this pattern for 20 years...

jkotas commented on src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs:
```diff
@@ -520,263 +522,235 @@ private unsafe string[] GetValueNamesCore(int values)
                         {
-                            // ERROR_MORE_DATA was returned however we cannot increase the buffer size beyond Int32.MaxValue
-                            Win32Error(r, name);
+                            Debug.Fail($"Expected non-negative data size, got {dataLength}.");
```
`RegQueryValueEx` returns `uint` (the PInvoke signature has mismatch), so this actually means that the data is more than 2GB. I think it would make more sense to throw ArithmeticException in this case. It is probably hypothetical case - I doubt that registry value can be more than 2GB.

stephentoub commented:
> Build breaks in CoreLib...

Ah, I'd searched for this file in *.csproj, forgot *.projitems.

jkotas reviewed (approved):
Nice!

jkotas commented on src/libraries/Common/src/Interop/Windows/Advapi32/Interop.RegQueryValueEx.cs:
```diff
@@ -14,39 +14,12 @@ internal static partial class Interop
+        internal static unsafe partial int RegQueryValueEx(
             SafeRegistryHandle hKey,
             string? lpValueName,
             int[]? lpReserved,
```
```suggestion
            int* lpReserved,
```
This is always null. We can make it pointer to make the marshaling cheaper.

danmoseley commented on src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs:
```diff
@@ -520,263 +523,198 @@ private unsafe string[] GetValueNamesCore(int values)
-                        char[] blob = new char[datasize / 2];
+                                    if (dataLength >= span.Length)
+                                    {
+                                        byte[] newPooled = ArrayPool<byte>.Shared.Rent(dataLength + 1);
```
is the rented array zeroed? as you're not necessarily copying over all of it.

stephentoub replied:
It's not zero'd. Where am I reading from an uninitialized portion of the array?

danmoseley replied:
I thought you were not writing to all of it. If your not reading the remainder then fine. I can't look at the code right now as I'm on my phone.

stephentoub replied:
It may not write to all of it, but it should only ever read from what's written.

danmoseley commented on src/libraries/System.Private.CoreLib/src/Internal/Win32/RegistryKey.cs:
```diff
@@ -7,6 +7,7 @@
+using System.Runtime.InteropServices;
 using System.Security;
 
 using Internal.Win32.SafeHandles;
```
there's no reasonable way to just use #ifdef to compile a subset of the full implementation, instead of copying part of the code? (I know this was always like this, not suggesting you change it here)

danmoseley commented:
@stephentoub are you still working on this one?

stephentoub commented:
> are you still working on this one?

Yes.

stephentoub committed:
Improve performance of RegistryKey.GetValue

Reduces typical number of syscalls as well as avoids allocation (or uses the ArrayPool otherwise).

Also updates Corelib's copy of RegistryKey to match (with things like perf key support removed).

stephentoub commented:
I believe I finally figured out the issues that were causing failures here; perf keys are special and have undocumented requirements (or at least requirements I couldn't find), like the input buffer has to be zero'd.

@jkotas, could you re-review?

jkotas reviewed (approved):
LGTM

stephentoub committed:
Merge branch 'main' into registryalloc
